doctor/lib/utils.py

```diff
@@ -354,3 +356,127 @@ def make_page_with_text(page, data, h, w):
     can.save()
     packet.seek(0)
     return packet
+
+
+def skew(obj):
+    """"""
+    if (matrix := obj.get("matrix")) is None:
+        return True
+
+    # Remove Skew
+    my_char_ctm = CTM(*matrix)
+    if my_char_ctm.skew_x != 0:
+        return False
+    return True
+
+
+def detect_words_per_page(document_text, page_count) -> float:
+    """Calculate word page frequency for the document
+
+    Generally - less than ten words is a suspect document for the length
+    of an entire document
+
+    :param document_text: the docuemnt text
+    :param page_count: number of pages
+    :return: average words per page
+    """
+    word_matches = re.findall(r"(\b([A-Z]?[a-z]){1,20}\b)", document_text)
+    list_of_words = list(set([matches[0] for matches in word_matches]))
+    return len(list_of_words) / page_count
+
+
+def page_has_text_annotations(path) -> bool:
+    """Page contains annotations
+
+    :param path: the path to the document
+    :return: True if free text or widget annotations in PDF
+    """
+    with pdfplumber.open(path) as pdf:
+        for page in pdf.pages:
+            if page.annots:
+                anno_types = [
+                    str(annot.get("data").get("Subtype"))
+                    for annot in page.annots
+                ]
+                if "/'FreeText'" in anno_types or "/'Widget'" in anno_types:
+                    return True
+    return False
+
+
+def get_page_text(page: pdfplumber.PDF.pages, strip_margin: bool):
+    """Extract page text
+
+    Using pdf plumber extract out the text of the document that is not
+    skewed (ie a stamp of approval) and extract out text removing blue text
+
+    :param page: pdf plumber page
+    :param strip_margin: a flag to crop out the margin of a document and skewed content
+    :return: text
+    """
+    if strip_margin:
+        # Crop margins and remove skewed text
+        bbox = (
+            (1 / 8.5 * page.width),
+            (1 / 11 * page.height),
+            (7.5 / 8.5 * page.width),
+            (10 / 11 * page.height),
+        )
+        doc_text = (
+            page.crop(bbox)
+            .filter(skew)
+            .extract_text(layout=True, keep_blank_chars=True)
+        )
+    else:
+        doc_text = page.extract_text(layout=True, keep_blank_chars=False)
+    return doc_text
+
+
+def extract_pdf_text(path: str, strip_margin: bool) -> str:
+    """Extract content from a PDF per page
+
+    :param path: Path to the pdf file
+    :param strip_margin: If this is a recap document to strip margins
+    :return: text on page
+    """
+    page_content = []
+    with pdfplumber.open(path) as pdf:
+        for page in pdf.pages:
+            text = get_page_text(page, strip_margin=strip_margin).strip()
+            page_content.append(text)
+    return "\n".join(page_content)
+
+
+def ocr_needed(path: str, content: str, page_count: int) -> [bool, Any]:
```

Comment:
Are the checks in this function in roughly performant order, such the fast checks come first, and difficult ones come later? Might be a good idea?